types/shares.go

```diff
@@ -184,3 +185,311 @@ func zeroPadIfNecessary(share []byte, width int) []byte {
+	if err != nil {
+		return MessagesEmpty, err
+	}
+
+	return Messages{
+		MessagesList: msgList,
+	}, nil
+}
+
+// processContiguousShares takes raw shares and extracts out transactions,
+// intermediate state roots, or evidence. The returned [][]byte do have
+// namespaces or length delimiters and are ready to be unmarshalled
+func processContiguousShares(shares [][]byte) (txs [][]byte, err error) {
+	if len(shares) == 0 {
+		return nil, err
+	}
+	share := shares[0][NamespaceSize+ShareReservedBytes:]
+	share, txLen, err := parseDelimiter(share)
+	if err != nil {
+		return nil, err
+	}
+
+	for i := 0; i < len(shares); i++ {
+		var newTxs [][]byte
+		newTxs, share, txLen, err = collectTxsFromShare(share, txLen)
+		if err != nil {
+			return nil, err
+		}
+
+		// add the collected txs to the output
+		txs = append(txs, newTxs...)
+
+		// if there is no next share, the work is done
+		// if not, then we know there are more shares to process
+		if len(shares) == i+1 {
+			break
+		}
+
+		nextShare := shares[i+1][NamespaceSize+ShareReservedBytes:]
+
+		// if there is no current share remaining, process the next share
+		if len(share) == 0 {
+			share, txLen, err = parseDelimiter(nextShare)
+			if err != nil {
+				break
+			}
+			continue
+		}
+
+		// create the next share by merging the next share with the remaining of
+		// the current share
+		share = append(share, nextShare...)
+	}
+
+	return txs, err
+}
+
+func collectTxsFromShare(share []byte, txLen uint64) (txs [][]byte, extra []byte, l uint64, err error) {
+	for uint64(len(share)) >= txLen {
+		tx := share[:txLen]
+		if len(tx) == 0 {
+			share = nil
+			break
+		}
+
+		txs = append(txs, tx)
+		share = share[txLen:]
+
+		share, txLen, err = parseDelimiter(share)
+		if txLen == 0 || err != nil {
+			break
+		}
+	}
+	return txs, share, txLen, err
+}
+
+// parseMessages iterates through raw shares and separates the contiguous chunks
+// of data. we use this for transactions, evidence, and intermediate state roots
+func parseMsgShares(shares [][]byte) ([]Message, error) {
+	if len(shares) == 0 {
+		return nil, nil
+	}
+
+	// set the first nid and current share
+	nid := shares[0][:NamespaceSize]
+	currentShare := shares[0][NamespaceSize:]
+
+	// find and remove the msg len delimiter
+	currentShare, msgLen, err := parseDelimiter(currentShare)
+	if err != nil {
+		return nil, err
+	}
+
+	var msgs []Message
+	for cursor := uint64(0); cursor < uint64(len(shares)); {
+		var msg Message
+		currentShare, nid, cursor, msgLen, msg, err = nextMsg(
+			shares,
+			currentShare,
+			nid,
+			cursor,
+			msgLen,
+		)
+		if err != nil {
+			return nil, err
+		}
+		if msg.Data != nil {
+			msgs = append(msgs, msg)
+		}
+	}
+
+	return msgs, nil
+}
+
+func nextMsg(shares [][]byte, current, nid []byte, cursor, l uint64) ([]byte, []byte, uint64, uint64, Message, error) {
+	switch {
+	// the message uses all of the current share data and at least some of the
+	// next share
+	case l > uint64(len(current)):
```

Comment:
What is the reason for the magic `1`? I'm not clear on what this code is doing.

Reply:
yikes, that is a terrible var name :sweat_smile: what about `msgLen`? https://github.com/lazyledger/lazyledger-core/pull/261/commits/8794411f54caf3ec87f6c3378d990b736ba6ffab